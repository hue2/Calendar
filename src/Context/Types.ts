export interface IEventDb {
    _id: string,
    note: string, 
    date: string,
    sticky: boolean,
}

export interface IContextProps {
    show: boolean,
    date: string,
    note: string,
    events: Array<IEvent>,
    importDest: string,
    sticky: boolean,
    stickyEvents: Array<IEvent>,
    onShow: () => void,
    onClose: () => void,
    onDatePick: (event : any) => void,
    onNoteSet: (event : any) => void,
    onEventSave: (id: string | null) => void,
    onEdit: (event : IEvent) => void,
    onDelete: () => void,
    onExportSelect: (data: any) => void,
    onExport: () => void,
    onImportSel: (data: string) => void,
    onImport: (event: any) => void;
    onSetSticky: (event: any) => void;
}

export interface IEvent {
    id: string,
    start: string,
    end: string,
    title: string,
    sticky: boolean
}
