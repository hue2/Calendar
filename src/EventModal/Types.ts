export interface EventModalProps {
    onShow: () => void,
    onHide: () => void,
    onDatePick: (event: any) => void,
    onNoteSet: (event: any) => void,
    onSave: () => void,
}