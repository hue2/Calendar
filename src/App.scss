@import '~@fullcalendar/core/main.css';
@import '~@fullcalendar/daygrid/main.css';
@import '~@fullcalendar/timegrid/main.css';
@import '~@fullcalendar/list/main.css';
@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@300;400&display=swap');
$dark-bg: #171e24;
$dark-blue: #353B47;
$light-blue: #306a87;

body {
  background: $dark-bg !important;
}

.App {
  text-align: right;
  padding: 15px;
  color: white;
  background: $dark-bg;
  font-family: 'Roboto', sans-serif;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

table tr, input {
  cursor: pointer;
}

table tr {
  &:hover {
    color: $dark-bg;
  }
}

.fc-view, .fc-unthemed .fc-list-empty {
  background-color: $dark-blue !important;
}

.fc-widget-header {
  background: #59596B !important;
}

.fc-view-container {
    border: none;
    div {
      border: none;
    }
}

.export-btn {
  font-size: 12px;
}

.f-left {
  float: left;
}

.fc-scroller {
  height: auto !important;
  overflow: auto;
}

.tab-content {
  margin-top: 15px;
}