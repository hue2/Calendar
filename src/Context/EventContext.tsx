import React, { useState, createContext, useEffect } from "react";

import Db from '../Database/Db';
import { IEventDb, IContextProps, IEvent } from './Types';

const fs = require('fs');

export const EventContext = createContext({} as IContextProps);

const db = new Db();

export function EventProvider(props: any) {
    let defaultDate = new Date().toISOString().substring(0, 10);

    const [show, setShow] = useState(false);
    const [date, setDate] = useState(defaultDate);
    const [note, setNote] = useState("");
    const [id, setId] = useState("");
    const [sticky, setSticky] = useState(false);
    const [events, setEvent] = useState<Array<IEvent>>([]);
    const [stickyEvents, setStickyEvents] = useState<Array<IEvent>>([]);

    useEffect(() => {
       getEvents();
    }, []);

    function handleShow() {
        setShow(true);
    }

    function getEvents() {
        db.getEvent().then((docs : any) => {
            let nonStickEvents = docs.filter(x => x.sticky === undefined || x.sticky === false);
            let allEvents = nonStickEvents.map((item : any) => ({
                id: item._id,
                title: item.note,
                start: item.date,
                end: item.end,
                sticky: false
            }));
            setEvent(allEvents);

            let stickEventsOnly = docs.filter(x => x.sticky !== undefined && x.sticky === true);
            let stickyEvents = stickEventsOnly.map((item : any) => ({
                id: item._id,
                title: item.note,
                start: item.date,
                end: item.end,
                sticky: true
            }));

            setStickyEvents(stickyEvents);
        });
    }

    function handleEditModalShow(event : IEvent) {
        setId(event.id);
        setNote(event.title);
        setDate(event.start);
        setSticky(event.sticky)
        setShow(true);
    }
    
    function handleClose() {
        handleReset();
        setShow(false);
    }

    function handleReset() {
        setId("");
        setNote("");;
        setDate(defaultDate);
        setSticky(false);
    }

    function handleDatePick(event : any) {
        setDate(event.target.value);
    }

    function handleNoteSet(event : any) {
        setNote(event.target.value);
    }
    
    function handleSetSticky(event : any) {
        setSticky(event.target.checked);
    }
    
    function handleEventSave() {
       if (id === null || id === "") {
            handleCreate();
       }
       else {
            handleEdit();
       }
    }

    function handleDelete() {
        db.deleteEvent(id).then(() => {
            handleClose();
            getEvents();
        })
    }

    function handleCreate() {
        db.createEvent({ date, note, sticky }).then((result : IEventDb) => {
            getEvents();
            handleClose();
        });
    }
     
    function handleEdit() {
        db.editEvent(id, { date, note, sticky }).then((result : IEventDb) => {
            getEvents();
            handleClose();
        });
    }  

    return (
        <EventContext.Provider value={{
                show, 
                date, 
                note,
                events,
                sticky,
                stickyEvents,
                onShow: handleShow, 
                onClose: handleClose,
                onDatePick: handleDatePick, 
                onNoteSet: handleNoteSet,
                onEventSave: handleEventSave,
                onEdit: handleEditModalShow,
                onDelete: handleDelete,
                onSetSticky: handleSetSticky,
            }}
        >
            {props.children}
        </EventContext.Provider>
    )
}
