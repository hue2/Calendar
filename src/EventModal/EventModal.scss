label {
    padding: 5px;
}

#date {
    border: 1px solid gray;
    border: 1px solid #ced4da;
    border-radius: 3px;
    padding: 3px;
}

#permanent {
    display: inline-block;
    width: 10%;
    height: 16px;
    padding-top: 10px;
}

.form-control {
    &:focus {
        border-color: none;
        box-shadow: none;
    }
}
