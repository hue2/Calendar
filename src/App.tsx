import React, { useContext } from 'react';

import FullCalendar from '@fullcalendar/react';
import listPlugin  from '@fullcalendar/list';
import dayGridPlugin  from '@fullcalendar/daygrid';
import interactionPlugin from '@fullcalendar/interaction';

import { EventModal } from './EventModal/EventModal';
import { EventContext } from './Context/EventContext';
import { IEvent } from './Context/Types';
import { StickyTable } from './StickyTable/StickyTable';
import './App.scss';

function App() {
  const context = useContext(EventContext);
  const { events, stickyEvents, onShow, onEdit } = context;

  function handleShow() {
    onShow();
  }

  function handleEditModal(event : any) {
    let data : IEvent = {
      id: event.id,
      title: event.title,
      start: event.start ? (new Date(event.start)).toISOString().substring(0, 10) : event.start,
      end: event.end,
      sticky: event.sticky,
    }
    onEdit(data);
  }

  return (
      <div className="App">
        <br />
       
            <FullCalendar 
              defaultView="listWeek" 
              views={{
                listWeek: { buttonText: 'Week' },
                listMonth: { buttonText: 'Month' },
                dayGrid30Day: {
                  type: 'listMonth',
                  duration: { days: 30 },
                  buttonText: 'Month'
                }
              }}
              header={{
                center: 'addEventButton',
                right: 'prev, listWeek, dayGrid30Day, next'
              }}
              plugins={[ interactionPlugin, listPlugin, dayGridPlugin  ]}
              selectable={true}
              customButtons={ 
                {
                  addEventButton: {
                    text: 'Add Event',
                    click: handleShow,
                  }
                }
              }
              allDayText={""}
              events={events}
              dateClick={(info) => {
                alert(info.dateStr);
              }}
              eventClick={(info) => {
                handleEditModal(info.event);
              }}
              eventColor='#E79B25'
            />
            <EventModal />
            <br />               
          <br/>
          <StickyTable stickyEvents={stickyEvents} handleEditModal={handleEditModal} />
       
      </div>
      
  );
}

export default App;
