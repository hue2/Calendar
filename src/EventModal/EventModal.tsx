import React, { useContext } from 'react';
import { Modal, Button, FormControl, Form } from 'react-bootstrap';

import { EventContext } from '../Context/EventContext';

import "react-datepicker/dist/react-datepicker.css";
import "./EventModal.scss";

export function EventModal() {
    const context = useContext(EventContext);
    const { show, date, note, onClose, onDatePick, onNoteSet, onEventSave, onDelete } = context;

    return (
        <Modal show={show} onHide={onClose}>
            <Modal.Header closeButton>
                <Modal.Title>
                    Add an event
                </Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <Form>
                    <Form.Group>
                        <Form.Label>
                            Date:
                        </Form.Label>
                        <FormControl id="date" type="date" 
                            value={date} 
                            onChange={onDatePick}
                        />

                        <br />
                        <Form.Label>
                            Note:
                        </Form.Label >
                        <FormControl id="note" type="text" 
                            onChange={onNoteSet}
                            value={note}
                            as="textarea"
                        />
                    </Form.Group>
                </Form>
            </Modal.Body>
            <Modal.Footer>
                <Button variant="danger" className="mr-auto" onClick={onDelete}>
                    Delete
                </Button>
                <Button variant="secondary" onClick={onClose}>
                    Close
                </Button>
                <Button variant="primary" onClick={() => onEventSave("hello")}>
                    Save
                </Button>
            </Modal.Footer>
        </Modal>
    )
}