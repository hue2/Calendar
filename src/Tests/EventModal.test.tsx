import React, { useContext } from 'react';
import { render, fireEvent, queryByAttribute  } from '@testing-library/react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow, mount } from 'enzyme';

import FullCalendar from '@fullcalendar/react';

import App from '../App';
import { EventContext } from '../Context/EventContext';
import { EventModal } from '../EventModal/EventModal';
import { EventProvider } from '../Context/EventContext';

Enzyme.configure({ adapter: new Adapter() });

describe("test event modal", () => {
    let wrapper, context, contextCallback;

    beforeEach(() => {
        context = {
            show: false,
            date: '2020/06/19',
            note: 'note note note',
            sticky: false, 
            onClose: jest.fn(),
            onDatePick: jest.fn(),
            onNoteSet: jest.fn(),
            onEventSave: jest.fn(),
            onDelete: jest.fn(),
            onSetSticky: jest.fn(),
        }
        contextCallback = jest.fn();

        wrapper = mount(
            <EventProvider value={context}>
                <EventModal />
                <EventContext.Consumer>
                    {contextCallback}
                </EventContext.Consumer>
            </EventProvider>
        );        
    });

    afterEach(() => {
        //wrapper.unmount();
    });

    it("should match snapshot", () => {
        const tree = shallow(<EventModal />);
        expect(tree).toMatchSnapshot();
    });

    it('should render', () => {
        jest.mock("../Context/");

        const { getByText } = render(
            <EventProvider value={context}>
                <EventModal />           
            </EventProvider>
        );

        const closeBtn = getByText('Close');
        fireEvent.click(closeBtn);
        expect(contextCallback).toHaveBeenCalledTimes(1);
    })
})