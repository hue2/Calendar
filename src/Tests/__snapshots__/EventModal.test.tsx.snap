// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test event modal should match snapshot 1`] = `ShallowWrapper {}`;
